lastlog2 - Y2038 safe version of lastlog

schubi2 opened the pull request:
* liblastlog2.so.0 - contains all high level functions to manage the data (util-linux/liblastlog2).
* pam_lastlog2.so - shows the last login of a user and stores the new login into the database (util-linux/pam_lastlog2).
* lastlog2 - will display the last logins for all users, who did ever login (util-linux/misc-utils).
* lastlog2-import.service - a service which imports lastlog data into lastlog2 database (util-linux/misc-utils).
* lastlog2.conf - configuration for tmpfiles.d. See tmpfiles.d(5) for details (util-linux/misc-utils).

For more information have a look to liblastlog2/README.md.

Requested by:
https://github.com/util-linux/util-linux/pull/2164#issuecomment-1689732782

schubi2 commented:
There are some missing sqlite3 in the buildsystem now. What has been expected :-) How to fix it ?

schubi2 commented on misc-utils/lastlog2.c:
```diff
@@ -0,0 +1,339 @@
+	if (tflg && ((time (NULL) - ll_time) > t_days))
+		return 0;
+
+	tm = localtime (&ll_time);
```
have moved to localtime_r

t-8ch commented:
> How to fix it ?

The way `readline` is handled in `configure.ac`, `Makefile.am` and `meson.build` is probably a good example.
Search for `READLINE_LIBS` in autotools files and `'readline'` in meson files.

I recommend to start with only one of the buildsystems and get it to work and reviewed.

thkukuk commented:
> > How to fix it ?
> 
> The way `readline` is handled in `configure.ac`, `Makefile.am` and `meson.build` is probably a good example. Search for `READLINE_LIBS` in autotools files and `'readline'` in meson files.
> 
> I recommend to start with only one of the buildsystems and get it to work and reviewed.

But this would mean that lastlog2 gets never tested, as you have to completely disable it in the build process for all CI tests.
Wouldn't it be better to add sqlite3-dev (or however the Ubuntu package name is) to the package list in `.github/workflows/cibuild-setup-ubuntu.sh`?

t-8ch commented:
> Wouldn't it be better to add sqlite3-dev

sqlite3 should already be pulled in transitively (it should also be added explicitly).
Currently the sqlite headers are already used and compilation itself works..

It then fails during linking because the makefiles just don't link with libsqlite.
Before merging both buildsystems should be finished.

t-8ch commented:
Could you also hook up the tests to the testsuite in `tests/ts/liblastlog2/...` ?

t-8ch commented on configure.ac:
```diff
@@ -1201,6 +1210,32 @@ AS_IF([test "x$build_libuuid" = xyes], [
 ])
 
+dnl
+dnl liblastlog2
```
This should default to autodetection instead of enabling by default.

schubi2 replied:
I have set the default value to enable_liblastlog2=check and checking if sqlite3 is available.

t-8ch commented on configure.ac:
```diff
@@ -1201,6 +1210,32 @@ AS_IF([test "x$build_libuuid" = xyes], [
 ])
 
+dnl
+dnl liblastlog2
```
Also add sqlite3 to `.packit.yaml` to test it there.

schubi2 replied:
I have adde sqlite-devel

t-8ch commented on tests/ts/liblastlog2/dlopen:
```diff
@@ -0,0 +1,15 @@
+
+TESTPROG="$TS_HELPER_LIBLASTLOG2_DELOPEN"
+
+[ -x $TESTPROG ] || ts_skip "test not compiled"
```
Use `ts_check_test_command`.

t-8ch commented on tests/ts/liblastlog2/dlopen:
```diff
@@ -0,0 +1,15 @@
+
+[ -x $TESTPROG ] || ts_skip "test not compiled"
+
+ts_run $TESTPROG >> $TS_OUTPUT.debug 2>&1
```
No need for `ts_run`.

t-8ch replied:
This testprog needs to be passed the `.so` as argument to actually do something.

t-8ch commented on tests/ts/liblastlog2/dlopen:
```diff
@@ -0,0 +1,15 @@
+. "$TS_TOPDIR"/functions.sh
+ts_init "$*"
+
+TESTPROG="$TS_HELPER_LIBLASTLOG2_DELOPEN"
```
typo in `DLOPEN`.

t-8ch commented on tests/ts/liblastlog2/pam_lastlog2_output:
```diff
@@ -0,0 +1,19 @@
+
+[ -x $TESTPROG ] || ts_skip "test not compiled"
+
+ts_run $TESTPROG >> $TS_OUTPUT.debug 2>&1
```
`$TS_OUTPUT` will be compared to a reference output to make sure that the tests executes as expected.
`$TS_OUTPUT.debug` is not checked in this way.

In general the tests as they are do not assert anything.
Try to work with "$TS_OUTPUT" and "$TS_ERRLOG" if possible and do manual checks with `ts_fail` if that's not sufficient.

t-8ch commented on tests/ts/liblastlog2/write_read_user:
```diff
@@ -0,0 +1,17 @@
+
+ts_run $TESTPROG >> $TS_OUTPUT.debug 2>&1
+
+rm "$TS_TOPDIR"/../tst-write-read-user.db
```
No need for the full path to the db file.
The helpers dump them into the current directory, so they can be referenced with it.

t-8ch commented on liblastlog2/src/tests/tst-y2038-64bit-time_t.c:
```diff
@@ -0,0 +1,43 @@
+main(void)
+{
+	if (sizeof (time_t) < 8) {
+		fprintf (stderr, "ERROR: time_t does not have at least 64bit\n");
```
And this *will* fail on i386.
Does the library really need 64bit times?

Ofc if a system is using 32bit times they won't be able to handle 64bit times properly but they should still be able to use liblastlog2, or?

thkukuk replied:
lastlog2 should solve the Y2038 problem on 64bit and 32bit systems. That's why we need a 64bit time_t even on 32bit architectures.
The 64bit time_t on 32bit is only used internal, so linking this with 32bit tools, which are not using 64bit time_t does not create any problems.
For util-linux your PR#2168, which enables -D_TIME_BITS=64 is needed. For the time being we can remove this test. This means, 32bit systems are not Y2038 safe even with lastlog2, but after util-linux enables 64bit time_t, it's safe again. And this change will not break the ABI, lastlog2 was designed from the beginning to be even Y2038 safe on systems with 32bit time_t.

t-8ch replied:
Unfortunately #2168 will break the externa ABI of `uudi_time()`.

In general I'm not sure about this test.
The library can be used without any reference to `time_t` at all.
So an application can use it on a platform with 32bit `time_t` through the explicit `int64_t` type and still be fine.

If an application wants to stuff the data into a 32bit `time_t` then that will break but it should still be allowed.
Maybe it's a 32bit application on a 64bit system.

We should also adapt all callers of liblastlog2 inside util-linux to use explicit `int64_t`.

schubi2 replied:
I have removed this test.

t-8ch commented on liblastlog2/man/ll2_read_all.3.adoc:
```diff
@@ -0,0 +1,68 @@
+const char *user = "root";
+
+static int
+callback (const char *res_user, time_t ll_time, const char *res_tty,
```
Above the `ll_time` parameter is `int64_t`, here it is `time_t`.

schubi2 replied:
Yes, I have had taken the example from lastlog2.c which has also been wrong and I have fixed afterword:
https://github.com/util-linux/util-linux/pull/2508/commits/87f8b2eb3f12c7cbbd68c03d9f2d131d2eac6ddf

t-8ch commented on liblastlog2/src/lastlog2.c:
```diff
@@ -0,0 +1,574 @@
+{
+	char *err_msg = NULL;
+	sqlite3_stmt *res;
+	char *sql_table = "CREATE TABLE IF NOT EXISTS Lastlog2(Name TEXT PRIMARY KEY, Time INTEGER, TTY TEXT, RemoteHost TEXT, Service TEXT) STRICT;";
```
The version of libsqlite used by Ubuntu 18.04 in the test suites fails with:
```
SQL error: near "STRICT": syntax error
```

schubi2 replied:
I have removed STRICT.

thkukuk commented:
> The library can be used without any reference to time_t at all.
> So an application can use it on a platform with 32bit time_t through the explicit int64_t type and still be fine.

That's correct for the application, but
> We should also adapt all callers of liblastlog2 inside util-linux to use explicit int64_t.

You cannot call the time functions with int64_t as argument, at least s390 would break horrible this way. That's why there are a few places in the code which use time_t and not int64_t. That's the only way to make sure that the time functions work correct on all architectures.
For the standalone lastlog2 the test was necessary to make sure, time_t is always 64bit independent of the architecture. But with the integration of util-linux the test doesn't make much sense, I agree. Or we build lastlog2 code with 64bit time_t even on 32bit architectures, but not the rest of util-linux.
But as written: for now we can live with a 32bit time_t and switch later to a 64bit time_t, this will neither break the ABI  nor the internal data structures.

t-8ch commented on tests/ts/liblastlog2/dlopen:
```diff
@@ -0,0 +1,13 @@
+
+ts_check_test_command $TS_HELPER_LIBLASTLOG2_DLOPEN
+
+$TS_HELPER_LIBLASTLOG2_DLOPEN "$TS_TOPDIR"/../.libs/liblastlog2.so >/dev/null || ts_failed "returned an error"
```
The test is failing during `distcheck`.
Unfortunately the output is discarded, so it's hard to know why.

It's weird that this test prints its failures to `stdout` in contrast to all the other tests which use `stderr`.

t-8ch commented:
> You cannot call the time functions

liblastlog2 is not calling any time functions, or?

> few places in the code which use time_t

These are currently failing to build on i386 exactly because it's passing mismatched pointers around.

> But as written: for now we can live with a 32bit time_t and switch later to a 64bit time_t, this will neither break the ABI nor the internal data structures.

Sounds good.

t-8ch commented on tests/ts/liblastlog2/dlopen:
```diff
@@ -6,10 +6,21 @@ TS_DESC="dlopen"
+else
+  #automake/autoconf build
+  libpath=${top_builddir}/.libs
+fi
```
This also won't work, the build directories can be anywhere with any name.

In `tests/functions.sh` we test for `$top_builddir/meson.conf`. That is probably the easiest thing for now.

t-8ch commented on liblastlog2/src/lastlog2.c:
```diff
@@ -0,0 +1,574 @@
+		if (error)
+			if (asprintf (error, "Cannot open database (%s): %s",
+				      path, sqlite3_errmsg (db)) < 0)
+				*error = strdup ("Out of memory");
```
It seems like a very unlikely edgecase that `asprintf` fails because of memory starvation but `strdup()` still succeeds.

schubi2 replied:
I have removed these strdups

t-8ch commented on liblastlog2/src/lastlog2.c:
```diff
@@ -0,0 +1,574 @@
+{
+	char *err_msg = NULL;
+	sqlite3_stmt *res;
+	char *sql_table = "CREATE TABLE IF NOT EXISTS Lastlog2(Name TEXT PRIMARY KEY, Time INTEGER, TTY TEXT, RemoteHost TEXT, Service TEXT);";
```
`static const`

schubi2 replied:
I have added.

t-8ch commented on liblastlog2/src/lastlog2.c:
```diff
@@ -0,0 +1,574 @@
+
+	sqlite3_close (db);
+
+	if (tty)
```
As mentioned before `free()` can handle `NULL` just fine everywhere.

schubi2 replied:
I have removed all these checks in this file.

t-8ch commented on liblastlog2/src/lastlog2.c:
```diff
@@ -0,0 +1,574 @@
+		for (int i = 0; i < argc; i++)
+			fprintf (stderr, " %s=%s", azColName[i], argv[i] ? argv[i] : "NULL");
+		fprintf (stderr, "\n");
+		exit (EXIT_FAILURE);
```
If this happens when called from PAM it will prevents users from logging in, no?

thkukuk replied:
If this prevents an user from logging in or not depends on the admin: he decides if errors from this module are fatal or not.
But calling "exit" here is fatal, this needs to be a return with an error value.

schubi2 replied:
The callback return -1 now and printing the error to stderr. sqlite3_exec is returning SQLITE_ABORT, which is returned by ll2_read_all too: "SQL error: query aborted". I think that should be enough because this error is more or less an internal error only, which should never happens or happens at once if someone changes the code in the wrong direction.
I have update the docu example to "optional" with the hint that it is up to the admin how this error should be handled (login is possible or not)

t-8ch commented on liblastlog2/src/lastlog2.h:
```diff
@@ -0,0 +1,60 @@
+
+#pragma once
+
+#define _PATH_LASTLOG2 "/var/lib/lastlog/lastlog2.db"
```
The leading underscore makes it seem like a private symbol.

`LL2_DEFAULT_DATABASE`?

Or remove it and let the users pass `NULL`  to fall back to the system-specific default.

t-8ch replied:
This comments doesn't seem to be addressed.

schubi2 replied:
This should be ok, or do I have something missunderstoud ?:
#define LL2_DEFAULT_DATABASE "/var/lib/lastlog/lastlog2.db"

t-8ch commented on liblastlog2/src/tests/tst-dlopen.c:
```diff
@@ -0,0 +1,43 @@
+
+#ifdef HAVE_CONFIG_H
+#  include <config.h>
+#endif
```
Not needed

schubi2 replied:
I have removed it.

t-8ch commented on liblastlog2/src/tests/tst-pam_lastlog2-output.sh:
Can be deleted

schubi2 replied:
I have removed it.

t-8ch reviewed (commented):
A few more comments.
Those can also be done after the merge if preferred.

Except the problem that the tests are not built under meson.

schubi2 commented on tests/ts/liblastlog2/dlopen:
```diff
@@ -0,0 +1,26 @@
+
+if [ -z "${top_builddir}" ]; then
+  # This should never happen
+  exit 1
```
Otherwise shellcheck would blame that this variable has not been initialized.
The other scripts in tests/ts have the same problem, but shellcheck will be called only if the source code has been
changed. So, is my way OK for you ?

t-8ch replied:
Shellcheck is wrong here. I would prefer to just drop the check.

schubi2 replied:
How can I do this ? It will be called in
https://github.com/util-linux/util-linux/actions/runs/6587935047/job/17899179605?pr=2508

t-8ch replied:
There will be a warning but it won't block the merge.

t-8ch commented on liblastlog2/src/lastlog2.h:
```diff
@@ -25,9 +25,7 @@
-#pragma once
-
-#define _PATH_LASTLOG2 "/var/lib/lastlog/lastlog2.db"
+#define LL2_DEFAULT_DATABASE "/var/lib/lastlog/lastlog2.db"
```
Did you think about letting callers just pass `NULL` to refer to the default location?
This avoids making the location part of the ABI.

schubi2 commented:
> Except the problem that the tests are not built under meson.
Yes indeed ! I have fixed it.

schubi2 commented:
Hi, 
may I ask if there is still something missed from my side :-)

t-8ch commented on misc-utils/lastlog2.c:
```diff
@@ -0,0 +1,339 @@
+				days = strtoul(optarg, &endptr, 10);
+				if ((errno == ERANGE && days == ULONG_MAX)
+				    || (endptr == optarg) || (*endptr != '\0')) {
+					fprintf (stderr, _("Invalid numeric argument: '%s'\n"), optarg);
```
`strtoul_or_err()`
(Also below)

t-8ch commented on liblastlog2/man/ll2_update_login_time.3.adoc:
```diff
@@ -0,0 +1,62 @@
+
+== RETURN VALUE
+
+Returns 0 on success, -1 on failure. _error_ contains an error string if the return value is -1.
```
It can also be `NULL` if the return value is `-1`.

schubi2 replied:
fixed

t-8ch commented on liblastlog2/src/lastlog2.h:
```diff
@@ -0,0 +1,60 @@
+  POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#pragma once
```
This comment is also not addressed.

t-8ch reviewed (commented):
There are still unaddressed old review comments.
Even though some are marked as resolved.

schubi2 commented:
> There are still unaddressed old review comments. Even though some are marked as resolved.

Ups yes, I have found them and will take care about it. Thanks !

schubi2 commented on liblastlog2/src/lastlog2.c:
```diff
@@ -0,0 +1,574 @@
+
+/* Check if database file exists.
+   Returns 0 on success, -1 on failure. */
+int ll2_check_database (const char *lastlog2_path)
```
That's a good idea. I have checked the DB open call of sqlite and they have had the same idea. ;-)
They are setting errno to ENOENT if the file does not exist.
So, I have removed this call or/and have replaced it by checking error.

schubi2 commented on liblastlog2/src/lastlog2.c:
```diff
@@ -0,0 +1,574 @@
+	if ((db = open_database_ro (lastlog2_path, error)) == NULL)
+		return -1;
+
+	char *sql = "SELECT * FROM Lastlog2 ORDER BY Name ASC";
```
I have fixed it.

schubi2 commented on liblastlog2/src/lastlog2.c:
```diff
@@ -0,0 +1,574 @@
+remove_entry (sqlite3 *db, const char *user, char **error)
+{
+	sqlite3_stmt *res;
+	char *sql = "DELETE FROM Lastlog2 WHERE Name = ?";
```
Also fixed and searched for others.....

schubi2 commented on liblastlog2/man/lastlog2.3.adoc:
```diff
@@ -0,0 +1,51 @@
+:man source: util-linux {release-version}
+:page-layout: base
+:lib: liblastlog2
+:firstversion: 2.39
```
Changed all to 2.40

schubi2 commented on liblastlog2/man/ll2_read_all.3.adoc:
```diff
@@ -0,0 +1,68 @@
+
+== RETURN VALUE
+
+Returns 0 on success, -1 on failure. _error_ contains an error string if the return value is -1.
```
I have added these lines.

schubi2 commented on liblastlog2/meson.build:
```diff
@@ -0,0 +1,37 @@
+    subdirs : 'lastlog2',
+    version : pc_version
+  )
+  if meson.version().version_compare('>=0.54.0')
```
I have removed it.

schubi2 commented on liblastlog2/src/tests/tst-remove-entry.c:
```diff
@@ -0,0 +1,88 @@
+		return 1;
+	}
+
+	if (error)
```
I have removed all these kind of checks in the testsuite.

schubi2 commented on pam_lastlog2/src/pam_lastlog2.c:
```diff
@@ -0,0 +1,343 @@
+ * or a pointer to the first char in STR after PREFIX.
+ */
+static inline const char *
+skip_prefix (const char *str, const char *prefix)
```
Using startswith.

schubi2 commented:
> Even though some are marked as resolved.
Yes, I have added a comment everytime and have had pushed the button "resolved". I have thought that my comment will be saved too but I have been wrong. 
I have added the comments again. Sorry for that.....

schubi2 commented:
> 
> There is one missing point regarding PAM and the exit in lastlog2.c. I am not sure there and I have asked Thorsten to have a look on it.

I have gotten an answer from Thorsten and have adapted the code.

t-8ch reviewed (approved):
Thanks for the patience!

As a last cleanup the commits could be squashed together as logically it is only one.

t-8ch commented:
For me it worked to rebase your branch on top of the latest master branch in the repository.
But it shouldn't matter either way.

karelzak commented:
> Is there anything else what I can do for this PR, or is there still an open point which I have had overseen ?

No, the ball is on my playground. I need to review it. Please, be patient-- my plan is to do it by the end of this year  ;-)

karelzak commented on configure.ac:
```diff
@@ -2577,6 +2612,13 @@ AS_IF([test "x$with_systemdsystemunitdir" != "xno"], [
+)
+AS_IF([test "x$with_tmpfilesdir" != "xno"], [
+  AC_SUBST([tmpfilesdir], [$with_tmpfilesdir])
+])
```
It seems we can ask PKG_CHECK_VAR() for the directory. See  `pkgconf systemd --variable=tmpfilesdir` output. We already use the same for systemdsystemunitdir, see AC_ARG_WITH([systemdsystemunitdir] block in the configure.ac.

karelzak replied:
This will be also usable for uuidd, now in Fedora / RHEL we have uuidd-tmpfiles.conf, I guess we can move it upstream.

schubi2 replied:
Wow, that's new for me. I have added it.

karelzak commented on liblastlog2/src/tests/tst_dlopen.c:
```diff
@@ -0,0 +1,39 @@
+/*
```
Please, add `SPDX-License-Identifier:` .

schubi2 replied:
"SPDX-License-Identifier: GPL-2.0-or-later" added

karelzak commented on misc-utils/lastlog2.c:
```diff
@@ -0,0 +1,332 @@
+{
+	FILE *output = (retval != EXIT_SUCCESS) ? stderr : stdout;
+
+	fprintf (output, _("Usage: lastlog2 [options]\n\nOptions:\n"));
```
Here, we need a little bit to convert it to the util-linux way :-)
```
fputs(USAGE_HEADER, out);
```
See, for example, the recently added misc-utils/waitpid.c.

schubi2 replied:
Yes, really nice. I have adapted it.

karelzak commented on misc-utils/lastlog2.c:
```diff
@@ -0,0 +1,332 @@
+	fputs (_("  -t, --time DAYS       Print only lastlog records more recent than DAYS\n"), output);
+	fputs (_("  -u, --user LOGIN      Print lastlog record of the specified LOGIN\n"), output);
+	fputs (_("  -v, --version         Print version number and exit\n"), output);
+	fputs ("\n", output);
```
```C
        fputs(USAGE_SEPARATOR, out);
        fprintf(out, USAGE_HELP_OPTIONS(25));
        
        fprintf(out, USAGE_MAN_TAIL("lastlog2(1)"));
```

schubi2 replied:
adapted. thanks !

karelzak commented on misc-utils/lastlog2.c:
```diff
@@ -0,0 +1,332 @@
+
+	if (argc > optind) {
+		fprintf (stderr, _("Unexpected argument: %s\n"), argv[optind]);
+		usage (EXIT_FAILURE);
```
We do not use usage() on error anymore in util-linux. Please print the error by err().

 (Or, in soma cases, we use warn() and errtryhelp(EXIT_FAILURE)).

karelzak replied:
I'll also suggest not using space between the function name and '('.

schubi2 replied:
I have used errx(). Is this OK ?

karelzak commented on liblastlog2/src/lastlog2.c:
```diff
@@ -0,0 +1,567 @@
+		if (tty) {
+			uc = sqlite3_column_text (res, 2);
+			if (uc != NULL && strlen ((const char *)uc) > 0)
+				*tty = strdup ((const char *)uc);
```
We need to care about strdup() success. What about
```
*tty = strdup ((const char *)uc);
if (!*tty)
  goto enomem;
...
enomem:
    return -1;  /* o return  -ENOMEM; */
```

You can use it also instead of `fprintf (stderr, "Out of memory\n");` for asprintf.

schubi2 replied:
I have added these "jumps". Looks better now :-)

karelzak commented on liblastlog2/src/lastlog2.c:
```diff
@@ -0,0 +1,567 @@
+
+			if (ll.ll_time != 0) {
+				int64_t ll_time;
+				char tty[UT_LINESIZE+1];
```
Maybe 
```
char tty[ sizeof(ll.ll_line) + 1];
```
Will be more robust.

schubi2 replied:
fixed

karelzak commented:
It looks good. I have some notes:

 I think the best design for APIs is to use some generic opaque handler. The reason is that such a library is possible to extend/change in arbitrary ways with minimal impact on the library's backward compatibility. 

(For example, I can imagine that one day, you decide on a different layout for the DB table or/and you will want to support multiple tables in the DB. You will be happy with a handler to change a table name for all ll2_ API functions.)

In the case of liblastlog2, `lastlog2_path` is everywhere as the first argument for API functions. I'd suggest using any `struct ll2_context` (or using any better name than `context`) rather than the path.
```
cxt = ll2_new_context("/path/to/db");
ll2_import_lastlog(cxt,  "/var/log/lastlog", &error);
ll2_unref_context(cxt);
```

You can also support NULL as context to get a default hardcoded behavior. ( `ll2_import_lastlog(NULL,  "/var/log/lastlog", &error);` ).

It seems the context would be usable, for example, as an application-defined pointer for sqlite3_exec() to generate an error message in the callback() on broken entries. Now it prints to stderr.

--
Please do not use fprintf(stderr)  in the shared library. It's fine to compose an error message and return it by &error pointer, but never write to stderr or stdout in the shared library.

schubi2 commented on liblastlog2/src/lastlog2.c:
```diff
@@ -0,0 +1,567 @@
+{
+	int retval = 0;
+	sqlite3_stmt *res;
+	static const char *sql = "SELECT * FROM Lastlog2 WHERE Name = ?";
```
Yes, indeed. Fixed.

schubi2 commented on liblastlog2/src/lastlog2.c:
```diff
@@ -0,0 +1,567 @@
+				char rhost[UT_HOSTSIZE+1];
+
+				ll_time = ll.ll_time;
+				strncpy (tty, ll.ll_line, UT_LINESIZE);
```
Nice ! Fixed.

schubi2 commented:
> cxt = ll2_new_context("/path/to/db");
> ll2_import_lastlog(cxt,  "/var/log/lastlog", &error);
> ll2_unref_context(cxt);

I like the name "context" :-) and have adapted your suggestion to the API.

> -- Please do not use fprintf(stderr) in the shared library. It's fine to compose an error message and return it by &error pointer, but never write to stderr or stdout in the shared library.

Yes, good point. I have removed all stdout,stderr calls.

schubi2 commented:
Hm, somehow three tests are failing:

fdisk/mbr-nondos-mode
fdisk/sunlabel
libfdisk/gpt

But I think that it is not an issue of this PR because other PRs have the same problems. e.g.:
https://github.com/util-linux/util-linux/pull/2647

So far I have tried to adapt all your request to the code. Thank you for your hints !
I have not rebased the code. Perhaps it helps you to see the changes faster.
Of course, I will have to rebase it again after you have checked the changes. :-)

karelzak commented:
Thanks! I have a few notes:

* It's better not to use typedef for structs. API is usually easier to use if you know that the type is struct. For more details see: https://www.kernel.org/doc/html/v4.10/process/coding-style.html#typedefs

* The optimal solution is to introduce lastlog2P.h (private library header file) and keep the struct definition there. It will also be easy to work with code in the future if you have public and private header files. The private file also makes keeping code in more .c files simple.

* Public lastlog2.h header file:
```C
struct ll2_context;
```
* Private lastlog2P.h header file:
```C
#include "lastlog2.h"

struct ll2_context {
    char *lastlog2_path;
};
```
* For the future extendability of the library, it would be better to be careful with `const` for the context struct. For example, in libmount, we almost nowhere use `const` for the context struct because, one day, it may be necessary to modify the struct for some reason in some of the API functions.

* Yes, ignore regression tests not related to lastlog2 ;-)

karelzak commented:
@schubi2 you need to add lastlog2P.h to liblastlog2/src/Makemodule.am:liblastlog2_la_SOURCES; otherwise, it will not be included in tarball (and `make distcheck` will not be happy).

The rest looks good.

@t-8ch any review & comment? ;-)

schubi2 commented:
> @schubi2 you need to add lastlog2P.h to liblastlog2/src/Makemodule.am:liblastlog2_la_SOURCES; otherwise, it will not be included in tarball (and `make distcheck` will not be happy).

Yes, thanks for the hint !

karelzak commented:
The wall.c problem should be fixed now; it's not your fault.

t-8ch reviewed (approved):
LGTM

t-8ch commented:
One last design discussion:
I'm not entirely sure about the context argument being optional.
It could make future changes to the library harder.

@schubi2 That last rebase seems to have gone wrong.

schubi2 commented:
> One last design discussion: I'm not entirely sure about the context argument being optional. It could make future changes to the library harder.

Hm, I like that way. But it is up to you what you are prefering. :-)

> 
> @schubi2 That last rebase seems to have gone wrong.

Yes, "git rebase" is sometimes tricky. At least for me ;-)

schubi2 commented:
Regarding the build error. Some other PRs have the same problem like:
https://github.com/util-linux/util-linux/pull/2711
So, I fear there is another reason for it than the current PR. Or I am wrong ?

t-8ch commented:
See https://github.com/util-linux/util-linux/pull/2718
Thanks for caring!
